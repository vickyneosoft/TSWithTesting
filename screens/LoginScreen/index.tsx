import React, {useCallback, useRef} from 'react';
import {NativeStackScreenProps} from '@react-navigation/native-stack/lib/typescript/src/types';
import {StyleSheet, View} from 'react-native';
import {useFormik} from 'formik';

// * Constants
import appConstants from '../../constants';
import colors from '../../constants/colors';

// * Components
import AppInput, {AppInputAPIs} from '../../components/AppInput';
import AppButton from '../../components/AppButton';
import BoldText from '../../components/BoldText';
import AuthFooter from '../../components/AuthFooter';
import AppLoader from '../../components/AppLoader';

// * HOC
import withSafeArea from '../../HOC/withSafeArea';
import withAuthContextConsumer from '../../HOC/withAuthContextConsumer';

// * Misc
import {ActionKind, AuthContextTypes, MyFormValues} from '../../types';
import {ErrorToast} from '../../utils/toastUtils';
// * - END of Imports Statements - *

const initialValues: MyFormValues = {email: '', password: ''};

const initialErrors: MyFormValues = {
  email: 'Email address required.',
  password: 'Password required',
};

const LoginScreen = (
  props: NativeStackScreenProps<any, any> & AuthContextTypes,
) => {
  const {navigation, authActionHandler, state} = props;
  const {isLoading} = state;

  const emailRef = useRef<AppInputAPIs>(null);
  const passwordRef = useRef<AppInputAPIs>(null);

  const onSignupRequestHandler = useCallback(() => {
    navigation.navigate('register');
  }, [navigation]);

  const validate = useCallback((values: MyFormValues) => {
    let hasErrors = true;
    const errors: MyFormValues = initialValues;

    if (!values.email) {
      errors.email = 'Email address required.';
    } else if (
      !/^[a-zA-Z0-9.+_-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,15}$/.test(values.email)
    ) {
      errors.email = 'Entered invalid email address.';
    } else {
      hasErrors = false;
    }

    if (!values.password) {
      errors.password = 'Password required';
    } else if (values.password.length < 6) {
      errors.password = 'Password should minimum of 6 characters';
    } else if (values.password.length > 15) {
      errors.password = 'Password should maximum of 15 characters';
    } else {
      hasErrors = false;
    }

    return hasErrors ? errors : undefined;
  }, []);

  const formSubmitHandler = useCallback(
    ({email, password}: MyFormValues) => {
      authActionHandler(ActionKind.LOGIN, {
        email,
        password,
      });
    },
    [authActionHandler],
  );

  const {errors, handleSubmit, handleChange} = useFormik({
    initialValues,
    initialErrors,
    onSubmit: formSubmitHandler,
    validate,
  });

  const onSubmitHandler = useCallback(() => {
    try {
      if (errors) {
        const errorMsgs = Object.keys(errors).filter(key => errors[key]);
        if (errorMsgs.length) {
          ErrorToast(errors[errorMsgs[0]]);
          return;
        }
      }
      handleSubmit();
    } catch (err: any) {
      ErrorToast(err?.message);
    }
  }, [errors, handleSubmit]);

  const onChangeTextHandler = useCallback(
    (inputId: string, _enteredText: string) => {
      switch (inputId) {
        case appConstants.EMAIL:
          handleChange('email')(_enteredText);
          break;
        case appConstants.PASSWORD:
          handleChange('password')(_enteredText);
          break;
      }
    },
    [handleChange],
  );

  const onSubmitEditingHandler = useCallback(
    (inputId: string) => {
      switch (inputId) {
        case appConstants.EMAIL:
          passwordRef.current?.focus();
          break;
        case appConstants.PASSWORD:
          onSubmitHandler();
          break;
      }
    },
    [onSubmitHandler],
  );

  return (
    <>
      <AppLoader isVisible={isLoading} />
      <View style={styles.container}>
        <View style={styles.container}>
          <View style={styles.placeholderContainer}>
            <BoldText style={styles.placeholderText}>{'Login'}</BoldText>
          </View>
          <AppInput
            ref={emailRef}
            placeholder="Email"
            keyboardType="email-address"
            onChangeText={onChangeTextHandler.bind(null, appConstants.EMAIL)}
            onSubmitEditing={onSubmitEditingHandler.bind(
              null,
              appConstants.EMAIL,
            )}
          />
          <AppInput
            ref={passwordRef}
            placeholder="Password"
            secureTextEntry
            blurOnSubmit={true}
            onChangeText={onChangeTextHandler.bind(null, appConstants.PASSWORD)}
            onSubmitEditing={onSubmitEditingHandler.bind(
              null,
              appConstants.PASSWORD,
            )}
          />
          <AppButton
            text="LOGIN"
            onPress={onSubmitHandler}
            style={styles.btn}
          />
        </View>
        <AuthFooter
          btnText="Sign Up"
          placeholderText={"Don't have an account? "}
          onPress={onSignupRequestHandler}
        />
      </View>
    </>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingHorizontal: 20,
    justifyContent: 'center',
    backgroundColor: colors.white,
  },
  placeholderContainer: {
    flex: 0.3,
    alignItems: 'center',
    justifyContent: 'center',
  },
  placeholderText: {alignSelf: 'center', fontSize: 22},
  btn: {
    marginTop: 20,
  },
  signupText: {
    flexShrink: 1,
    alignSelf: 'center',
    marginBottom: 10,
    fontSize: 14,
  },
});

export default withSafeArea(withAuthContextConsumer(LoginScreen));
