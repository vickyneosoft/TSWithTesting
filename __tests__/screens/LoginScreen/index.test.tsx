import React from 'react';
import LoginScreen from '../../../screens/LoginScreen';

// Note: test renderer must be required after react-native.
import renderer, { act } from 'react-test-renderer';

describe('react native firebase auth implementation test', () => {
    const firebaseAuth = require('@react-native-firebase/auth')

    it('should login with valid credentials', async () => {
        const res = await firebaseAuth.signInWithEmailAndPassword('abc@gmail.com', '123456')
        expect(res).toBe('success')
    })

    it('should throw error on invalid email address', async () => {
        try {
            await firebaseAuth.signInWithEmailAndPassword('abc', '123456')
        } catch (err: any) {
            expect(err).toBe('Invalid email-address')
        }
    })

    it('should throw error on less then 6 characters password length', async () => {
        try {
            await firebaseAuth.signInWithEmailAndPassword('abc@gmail.com', '123')
        } catch (err: any) {
            expect(err).toBe('Password should minimum of 6 characters')
        }
    })

    it('should throw error on greater then 15 characters password length', async () => {
        try {
            await firebaseAuth.signInWithEmailAndPassword('abc@gmail.com', '123456789101010')
        } catch (err: any) {
            expect(err).toBe('Password should maximum of 15 characters')
        }
    })
})

describe('<LoginScreen /> rendering', () => {
    it('LoginScreen screen renders correctly', () => {
        const screen = renderer.create(<LoginScreen />);
        const tree = screen.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('LoginScreen screen renders correctly', async () => {
        const screen = renderer.create(<LoginScreen />);
        const emailInput = screen.root.findByProps({
            testID: 'txtEmail'
        })
        const passwordInput = screen.root.findByProps({
            testID: 'txtPassword'
        })

        const submitBtn = screen.root.findByProps({
            testID: 'btnSubmit'
        })
        await act(async () => {
            emailInput?._fiber?.ref?.current?.setText('abcd@gmail.com')
            passwordInput?._fiber?.ref?.current?.setText('123456')
            await submitBtn?._fiber?.memoizedProps?.onPress()
        })
        expect(screen.toJSON()).toMatchSnapshot();
    });
});