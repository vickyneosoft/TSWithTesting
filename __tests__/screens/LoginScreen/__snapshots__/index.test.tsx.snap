// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> rendering LoginScreen screen renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flexGrow": 1,
    }
  }
>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    presentationStyle="overFullScreen"
    statusBarTranslucent={true}
    transparent={true}
    visible={false}
  />
  <View
    style={
      {
        "backgroundColor": "#343434",
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#343434",
          "flex": 1,
          "justifyContent": "center",
          "paddingHorizontal": 20,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 0.3,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 14,
                "fontWeight": "bold",
              },
              {
                "color": "#ffffff",
                "fontSize": 22,
              },
            ]
          }
        >
          Login
        </Text>
        <Text
          style={
            [
              {
                "fontSize": 12,
              },
              {
                "color": "#ffffff",
              },
            ]
          }
        >
          Enter your credentials to access account
        </Text>
      </View>
      <View
        style={
          {
            "marginVertical": 10,
          }
        }
      >
        <TextInput
          blurOnSubmit={false}
          keyboardType="email-address"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Email"
          placeholderTextColor="#ffffff"
          returnKeyType="next"
          style={
            {
              "borderColor": "#ffffff",
              "borderRadius": 30,
              "borderWidth": 2,
              "color": "#ffffff",
              "height": 55,
              "overflow": "hidden",
              "paddingLeft": 20,
            }
          }
          testID="txtEmail"
          value=""
        />
      </View>
      <View
        style={
          {
            "marginVertical": 10,
          }
        }
      >
        <TextInput
          blurOnSubmit={true}
          keyboardType="default"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Password"
          placeholderTextColor="#ffffff"
          returnKeyType="next"
          secureTextEntry={true}
          style={
            {
              "borderColor": "#ffffff",
              "borderRadius": 30,
              "borderWidth": 2,
              "color": "#ffffff",
              "height": 55,
              "overflow": "hidden",
              "paddingLeft": 20,
            }
          }
          testID="txtPassword"
          value=""
        />
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "borderRadius": 30,
              "borderWidth": 0.5,
              "height": 55,
              "justifyContent": "center",
            },
            {
              "marginTop": 20,
            },
          ]
        }
        testID="btnSubmit"
      >
        <Text
          style={
            {
              "fontSize": 14,
              "fontWeight": "bold",
            }
          }
        >
          LOGIN
        </Text>
      </View>
    </View>
    <Text
      onPress={[Function]}
      style={
        [
          {
            "fontSize": 12,
          },
          [
            {
              "color": "#ffffff",
              "fontSize": 16,
            },
            {
              "alignSelf": "center",
              "flexShrink": 1,
              "fontSize": 14,
              "marginBottom": 10,
            },
          ],
        ]
      }
      testID="btnSignup"
    >
      Don't have an account? 
      <Text
        style={
          [
            {
              "fontSize": 14,
              "fontWeight": "bold",
            },
            {
              "textDecorationLine": "underline",
            },
          ]
        }
      >
        Sign Up
      </Text>
    </Text>
  </View>
</View>
`;

exports[`<LoginScreen /> rendering LoginScreen screen renders correctly 2`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flexGrow": 1,
    }
  }
>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    presentationStyle="overFullScreen"
    statusBarTranslucent={true}
    transparent={true}
    visible={false}
  />
  <View
    style={
      {
        "backgroundColor": "#343434",
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#343434",
          "flex": 1,
          "justifyContent": "center",
          "paddingHorizontal": 20,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 0.3,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 14,
                "fontWeight": "bold",
              },
              {
                "color": "#ffffff",
                "fontSize": 22,
              },
            ]
          }
        >
          Login
        </Text>
        <Text
          style={
            [
              {
                "fontSize": 12,
              },
              {
                "color": "#ffffff",
              },
            ]
          }
        >
          Enter your credentials to access account
        </Text>
      </View>
      <View
        style={
          {
            "marginVertical": 10,
          }
        }
      >
        <TextInput
          blurOnSubmit={false}
          errorMsg="Email address required."
          keyboardType="email-address"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Email"
          placeholderTextColor="#ffffff"
          returnKeyType="next"
          style={
            {
              "borderColor": "#ffffff",
              "borderRadius": 30,
              "borderWidth": 2,
              "color": "#ffffff",
              "height": 55,
              "overflow": "hidden",
              "paddingLeft": 20,
            }
          }
          testID="txtEmail"
          value="abcd@gmail.com"
        />
        <Text
          style={
            [
              {
                "fontSize": 12,
              },
              {
                "color": "red",
                "marginTop": 2,
                "paddingLeft": 20,
              },
            ]
          }
        >
          Email address required.
        </Text>
      </View>
      <View
        style={
          {
            "marginVertical": 10,
          }
        }
      >
        <TextInput
          blurOnSubmit={true}
          errorMsg="Password required."
          keyboardType="default"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Password"
          placeholderTextColor="#ffffff"
          returnKeyType="next"
          secureTextEntry={true}
          style={
            {
              "borderColor": "#ffffff",
              "borderRadius": 30,
              "borderWidth": 2,
              "color": "#ffffff",
              "height": 55,
              "overflow": "hidden",
              "paddingLeft": 20,
            }
          }
          testID="txtPassword"
          value="123456"
        />
        <Text
          style={
            [
              {
                "fontSize": 12,
              },
              {
                "color": "red",
                "marginTop": 2,
                "paddingLeft": 20,
              },
            ]
          }
        >
          Password required.
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "borderRadius": 30,
              "borderWidth": 0.5,
              "height": 55,
              "justifyContent": "center",
            },
            {
              "marginTop": 20,
            },
          ]
        }
        testID="btnSubmit"
      >
        <Text
          style={
            {
              "fontSize": 14,
              "fontWeight": "bold",
            }
          }
        >
          LOGIN
        </Text>
      </View>
    </View>
    <Text
      onPress={[Function]}
      style={
        [
          {
            "fontSize": 12,
          },
          [
            {
              "color": "#ffffff",
              "fontSize": 16,
            },
            {
              "alignSelf": "center",
              "flexShrink": 1,
              "fontSize": 14,
              "marginBottom": 10,
            },
          ],
        ]
      }
      testID="btnSignup"
    >
      Don't have an account? 
      <Text
        style={
          [
            {
              "fontSize": 14,
              "fontWeight": "bold",
            },
            {
              "textDecorationLine": "underline",
            },
          ]
        }
      >
        Sign Up
      </Text>
    </Text>
  </View>
</View>
`;
